File: Army(java)/Basic/Health.java
package Basic;

public class Health {
    protected int damage;
    protected int hitPoints;
    protected int hitPointsLimit;

    public Health(int hp, int dmg) {
        this.hitPoints = hp;
        this.hitPointsLimit = hp;
        this.damage = dmg;
    }
        
    public void ensureIsAlive() throws UnitIsDeadException {
        if ( hitPoints == 0 ) {
            throw new UnitIsDeadException();
        }
    }

    public int getDamage() {
        return damage;
    }

    public void setDamage(int dmg) {
        damage = dmg;
    }

    public int getHitPoints() {
        return hitPoints;
    }

    public void setHitPoints(int hp) {
        hitPoints = hp;
    }

    public int getHitPointsLimit() {
        return hitPointsLimit;
    }

    public void setHitPointsLimit(int hp) {
        hitPointsLimit = hp;
    }

    public void addHitPoints(int hp) throws UnitIsDeadException {
        ensureIsAlive();
        
        if ( hp <= (hitPointsLimit - hitPoints) ) {
            hitPoints += hp;
        } else {
            hitPoints = hitPointsLimit;
        }
    }

    public void takeDamage(int dmg) throws UnitIsDeadException {
        ensureIsAlive();
        
        hitPoints -= dmg;
        
        if ( hitPoints < 0 ) {
            hitPoints = 0;
        }
    }
    
    public void printHealth() {
        System.out.println("Dagame\t\t" + " " + damage);
        System.out.println("Hit points\t" + " " + hitPoints);
        System.out.println("Hit points limit " + hitPointsLimit);
    }
}

File: Army(c++)/Basic/Health.cpp
#include "Health.hpp"

Health::Health(int hp, int dmg) : hitPoints(hp), hitPointsLimit(hp), damage(dmg) {}

Health::~Health() {}

void Health::ensureIsAlive() {
    if ( this->hitPoints == 0 ) {
       throw UnitIsDead();
   }
}

int Health::getDamage() const {
    return this->damage;
}

void Health::setDamage(int dmg) {
    this->damage = dmg;
}

int Health::getHitPoints() const {
    return this->hitPoints;
}

void Health::setHitPoints(int hp) {
    this->hitPoints = hp;
}

int Health::getHitPointsLimit() const {
    return this->hitPointsLimit;
}

void Health::setHitPointsLimit(int hp) {
    this->hitPointsLimit = hp;
}

void Health::addHitPoints(int hp) {
    this->ensureIsAlive();
    
    if ( hp <= (this->hitPointsLimit - this->hitPoints) ) {
        this->hitPoints += hp;
    } else {
        this->hitPoints = this->hitPointsLimit;
    }
}

void Health::takeDamage(int dmg) {
    this->ensureIsAlive();
    
    this->hitPoints -= dmg;
    
    if ( this->hitPoints < 0 ) {
        this->hitPoints = 0;
    }
}

std::ostream& operator<<(std::ostream& out, const Health& health) {
    out << "Damage " << health.getDamage() << std::endl;
    out << "Hit points " << health.getHitPoints() << std::endl;
    out << "Hit points limit " << health.getHitPointsLimit() << std::endl;
    
    return out;
}
